systemd/escape_test.go

```diff
@@ -22,16 +22,123 @@ package systemd_test
 import (
 	. "gopkg.in/check.v1"
 
-	. "github.com/snapcore/snapd/systemd"
+	"github.com/snapcore/snapd/systemd"
 )
 
-func (s *SystemdTestSuite) TestEscape(c *C) {
-	c.Check(EscapeUnitNamePath("Hallöchen, Meister"), Equals, `Hall\xc3\xb6chen\x2c\x20Meister`)
+func (ts *SystemdTestSuite) TestEscapePath(c *C) {
+	tt := []struct {
+		in      string
+		out     string
+		comment string
+	}{
+		{
+			`Hallöchen, Meister`,
+			`Hall\xc3\xb6chen\x2c\x20Meister`,
+			`utf-8 char and spaces`,
+		},
+		{
+			`/tmp//waldi/foobar/`,
+			`tmp-waldi-foobar`,
+			`unclean path`,
+		},
+		{
+			`/.foo/.bar`,
+			`\x2efoo-.bar`,
+			`leading dot escaped differently`,
+		},
+		{
+			`.foo/.bar`,
+			`\x2efoo-.bar`,
+			`leading dot escaped differently (without leading slash)`,
+		},
+		{
+			``,
+			`-`,
+			`empty string`,
+		},
+		{
+			`/`,
+			`-`,
+			`root /`,
+		},
+		{
+			`////////`,
+			`-`,
+			`root / unclean`,
+		},
+		{
+			`-`,
+			`\x2d`,
+			`just dash`,
+		},
+		{
+			`/run/ubuntu------data`,
+			`run-ubuntu\x2d\x2d\x2d\x2d\x2d\x2ddata`,
+			`consecutive dashes`,
+		},
+		{
+			`/path-with-forward-slash\`,
+			`path\x2dwith\x2dforward\x2dslash\x5c`,
+			`forward slash in path element`,
+		},
+		{
+			`/run/1000/numbers`,
+			`run-1000-numbers`,
+			`numbers`,
+		},
+		{
+			`/silly/path/,!@#$%^&*()`,
+			`silly-path-\x2c\x21\x40\x23\x24\x25\x5e\x26\x2a\x28\x29`,
+			`ascii punctuation`,
+		},
+		{
+			`/run/mnt/data`,
+			`run-mnt-data`,
+			`typical data initramfs mount`,
+		},
+		{
+			`run/mnt/data`,
+			`run-mnt-data`,
+			`typical data initramfs mount w/o leading slash`,
+		},
+		{
+			`/run/mnt/ubuntu-seed`,
+			`run-mnt-ubuntu\x2dseed`,
+			`typical ubuntu-seed initramfs mount`,
+		},
+		{
+			`/run/mnt/ubuntu data`,
+			`run-mnt-ubuntu\x20data`,
+			`path with space in it`,
+		},
+		{
+			`/run/mnt/ubuntu_data`,
+			`run-mnt-ubuntu_data`,
+			`path with underscore in it`,
+		},
+		{
+			` `,
+			`\x20`,
+			"space character",
+		},
+		{
+			`	`,
+			`\x09`,
+			`tab character`,
+		},
+		{
+			`/home/日本語`,
+			`home-\xe6\x97\xa5\xe6\x9c\xac\xe8\xaa\x9e`,
+			`utf-8 characters`,
+		},
+		{
+			`/path⌘place-of-interest/hello`,
+			`path\xe2\x8c\x98place\x2dof\x2dinterest-hello`,
+			`utf-8 character embedded in ascii path element`,
+		},
```

Comment:
Perhaps we could also test `/a/../b` which is invalid for `systemd-escape --path`? There's no "right" conversion here but we can at least make sure we don't crash and the result is deterministic.

Reply:
ah good catch, I wasn't aware that there were such cases, but yes we should probably do something like that. Ideally we would do the same and not encode them either with an error, but it's not clear to me how systemd decides whether a filepath is invalid or not, at least I don't see anything in the man page about it